### ISSUE TITLE
Improve accuracy when parameter is an integer 

### ISSUE THREAD

**Issue by radnado** [fmfn/BayesianOptimization#63]
The BO maximize method generates parameter values for iterations as type float.  This is expected and is fine.  However some functions (e.g. xgboost) have parameters that need to be an integer.  The current conversion practice is for the function to cast the parameter to an integer e.g. int(max_depth).  This will cast a float of say 4.8 as an integer of 4. 

The issue is that the score was calculated with a parameter of 4, but when returned to BO, BO uses it in future calculations as if it was calculated at 4.8.  My thinking is that this will lead to inaccuracies and degradation of the optimization process. 

This could possibly be solved by having the function also return all the parameter values that were actually used for the calculation so that BO could update the parameters for that iteration, enabling BO to use the correct parameter value (the integer) when calculating the parameters for the next iteration.

This would accommodate any changes (rounding or otherwise) that need to be made to the parameters in the function while protecting the accuracy of the optimization calculations. 

1. What are you thoughts on this as a long-term solution?

2. Your site examples cast float to integers as per int(x).  This does a truncation (e.g. 4.8 to 4) and might lead to inaccuracies.  I've used int(round(x)) to reduce the error in casting to an integer.  If you think it appropriate the examples would be updated to use this approach.  

3. In an attempt to get around this issue in the interim, I am processing the iterations in small batches of 2 iterations (e.g. bayesOpt.maximize(n_iter=2, acq="ei", xi=0.1) and then rounding the integer parameter values in bayesOpt.res['all']['params'].  I am hoping that by updating this dictionary with the integer values that were actually used to produce the score, that these correct values will then be used by BO when it calculates the next parameters.  Will updating this dictionary this way have this effect?

4. For the initialization observations (init_points) it might be appropriate to treat the parameters in the same way.  i.e. have all actual parameters used returned from the function and the relevant BO dictionary updated ready for optimization processing.  

This issue might be related to and resolve some of the discussions in issue #13. 
I like the package and hope this is useful. 
cheers

**Comment by fmfn** [comment]
This is and probably will always arguably be the most uncomfortable aspect of this method --- at least in the way it was implemented here. Just as a disclaimer, the goal of this package was to be simple and transparent, and to offer a fairly out of the box, plug and play, solution for BO. So I often made the conscious decision of leaving features and / or methods out for the sake of keeping it lean, and letting those who need more use packages such as spearmint.

Now to address your points:

1. One serious problem with limiting the parameter space BO has access to (which is the case when you don't allow it to probe 4.8) is that you may fall into an infinite loop --- I've been there. Say BO just probed 4.0, now it picks 4.2 as its next guess, if you force the value of 4.2 to 4.0 internally, BO will ask for 4.2 again on the next step, after all, its internal state did not change, all it knows is the value of 4.0 and 4.2 will still seem like a good choice.
Moreover this doesn't address the real problem that the underlying gaussian process, and hence the BO process, operate on a continuous space, while some parameters of the function being optimized are discrete.

2. Using round(x) is just as good of a choice, and it may make the process a bit more optimal, I haven't really tested one against the other to be honest. However I don't really expect big changes, the function being optimized is still C^0 (continuous but not differentiable everywhere), you're just changing where the step occurs (4.5, 5.5, ..., instead of 4.0, 5.0, ...).

3. Yes, updating `bayesOpt.res['all']['params']` will altear what the optimizer knows about your target function and it will accomplish what you are trying to do.

4. I agree that the entire discussion applies to `init_points` as well.

I understand that casting floats to ints (whether you round them or not) seems like a gross approximation, one that might even hurt the optimization process. Nonetheless I believe it is not that big of an issue provided it is addressed properly by setting the GP parameters adequately. Look at this function:
![download](https://user-images.githubusercontent.com/7664252/30749580-d5a3b634-9f81-11e7-8a7e-d051cbe8393a.png)
This is kinda of what you get when you set `int(x)` as is often done around here. It is definitely possible to approximate this function with a gaussian process, however in practice the choice of kernel and noise will be important, and may even determine if you're successful or not. 

Ultimately I rather pick a more flexible Gaussian Process (higher `alpha` and a kernel such as Mattern) and let BO do its thing, than try to restrict it

**Comment by radnado** [comment]
Thanks for your response fernando.  It really helps me understand the problem; I didn't think of the possible loop situation.  
It a great package (thank you!) and is very effective so I will continue on.  
Thanks again. cheers.

**Comment by lwaldner** [comment]
Hi I have a similar use case as @radnado and after reading through this I couldn't find any references to the "Mattern" kernel on sklearn or really anywhere so I was hoping someone could point me in the right direction. Also i'm using an alpha of 1e-5 but I was hoping to get some help on getting a good range of 'aggressive' alpha values. Thanks!

**Comment by fmfn** [comment]
http://scikit-learn.org/stable/modules/gaussian_process.html#matern-kernel
